scripts/portal-fetcher/openshift-docs-downloader.py

```diff
@@ -0,0 +1,1384 @@
+        url TEXT,
+        local_path TEXT,
+        status TEXT,
+        etag TEXT,
+        last_modified TEXT,
+        timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
+        change_type TEXT
+    )
+    """
+    )
+
+    conn.commit()
+    conn.close()
+    logger.info(f"Database initialized at {db_path}")
+
+    return db_path
+
+
+def record_download(
+    db_path: str,
+    url: str,
+    local_path: str,
+    status: str = "success",
+    etag: Optional[str] = None,
+    last_modified: Optional[str] = None,
+    change_type: Optional[str] = None,
+) -> str:
+    """Record a download in the database and update history
+
+    Args:
+        db_path (Path): Path to SQLite database
+        url (str): URL of the downloaded page
+        local_path (str/Path): Local path where the file is saved
+        status (str): Download status (success, failed, error)
+        etag (str): ETag header from the response, if available
+        last_modified (str): Last-Modified header from the response, if available
+        change_type (str): Type of change (new, updated, unchanged, error)
+    """
+    conn = sqlite3.connect(db_path)
+    cursor = conn.cursor()
+    local_path_str = str(local_path)
+
+    try:
+        # Record in main downloads table
+        cursor.execute(
+            "INSERT OR REPLACE INTO downloads (url, local_path, status, etag, last_modified, timestamp) VALUES (?, ?, ?, ?, ?, datetime('now'))",
+            (url, local_path_str, status, etag, last_modified),
+        )
+
+        # Record in history table
+        if change_type:
+            cursor.execute(
+                "INSERT INTO download_history (url, local_path, status, etag, last_modified, change_type) VALUES (?, ?, ?, ?, ?, ?)",
+                (url, local_path_str, status, etag, last_modified, change_type),
+            )
+
+        conn.commit()
+    except Exception as e:
+        logger.error(f"Database error: {e}")
+    finally:
+        conn.close()
+
+    return url
+
+
+def get_download_status(db_path: str, url: str) -> dict:
+    """Get download status from database
+
+    Args:
+        db_path (Path): Path to SQLite database
+        url (str): URL to check
+
+    Returns:
+        tuple: (etag, last_modified)
+    """
+    conn = sqlite3.connect(db_path)
+    cursor = conn.cursor()
+
+    cursor.execute(
+        "SELECT etag, last_modified FROM downloads WHERE url = ? AND status = 'success'",
+        (url,),
+    )
+    result = cursor.fetchone()
+
+    conn.close()
+
+    existing_etag = result[0] if result else None
+    existing_last_modified = result[1] if result else None
+
+    return (existing_etag, existing_last_modified)
+
+
+def get_download_results(db_path: str) -> tuple[set[str], set[str]]:
+    """Get all successful and failed downloads
+
+    Args:
+        db_path (Path): Path to SQLite database
+
+    Returns:
+        tuple: (successful_urls, failed_urls)
+    """
+    conn = sqlite3.connect(db_path)
```

Comment:
it theory you can do:

```python
with sqllite3.connect(db_path) as conn:
    ...
```

to avoid the need to call `close` at the end (which won't work in case of any exception)